crates/env/src/event.rs

```diff
@@ -215,3 +209,16 @@ pub trait Event: scale::Encode {
         E: Environment,
         B: TopicsBuilderBackend<E>;
 }
+
+/// Getter that returns the signature topic for the specific event.
+///
+/// It can be automatically calculated or manually specified.
+///
+/// The unique signature topic of the event. `None` for anonymous events.
+///
+/// Usually this is calculated using the `#[derive(ink::Event)]` derive, which by
+/// default calculates this as `blake2b("Event(field1_type,field2_type)")`
+pub trait GetSignatureTopic {
```

Comment:
I'm not convinced of the need for this additional trait, it is not useful on its own.

Reply:
It is useful when devs want to specify custom logic for implementing the generation of the topic signature (e.g. using a different hashing scheme or prepending it with some index). They can implement only the `GetSignatureTopic` trait instead of the whole `Event` trait.

It is also hard to support both otherwise `derive(Event)` proc macro, and nested `ink(event)` attr with the configuration options. 

The alternative would be to eliminate individual derive macros in favour of a single `#[ink::event]` proc macro that would allow the generation of the custom signature topic "in-place".

The other approach is to duplicate logic, and parse nested `#ink(signature_topic)` inside `event_derive`, similar to how it's done with `anonymous`.

Reply:
> It is useful when devs want to specify custom logic for the generation of the topic signature (e.g. using a different hashing scheme or prepending it with some index). They can implement only the GetSignatureTopic trait instead of the whole Event trait.

As per your example in the description:
```
impl GetSignatureTopic for Flipper {
   const SIGNATURE_TOPIC: Option<[u8; 32]> = <literal value or const fn>
}
```
I don't see how this adds anything over and above specifying it directly in the `signature_topic = <literal>`. It is a `const` so in all cases it is calculated at compile time. Sure it is a bit more complicated to add a `const fn` there because of attribute parsing, but I don't think we need to necessarily support that to start with, the user can just perform the equivalent in a custom macro if they really want.

> It is also hard to support both otherwise derive(Event) proc macro, and nested ink(event) attr with the configuration options.

I don't think we need to add anything to the legacy `#[ink(event)]` inline macro, I left that there for backwards compatibility, if they want to `signature_topic` they can just switch to `ink::event`. 

> Nonetheless, I am in favour of multiple small traits serving specific logic. GetSignatureTopic generates the signature topic in the event generation pipeline and, therefore, is self-sufficient. 

I agree with the principle of having multiple smaller traits where clear responsibilities can be separated. In this case, however, they coherently belong together. The trait was originally called `Topics` (which I renamed to `Event` (mistakenly?) because it is responsible for the topics of an event, and as part of that I moved the signature topic from being returned from `topics` to a `const`.

> Since we want to allow developers to customise the topic, we want to allow them to customise its generation as well

I argue above that with the `signature_topic = <literal>`  that they are able to customise it's generation, without having to understand about underlying traits.

Also be mindful that these traits are a public API and we want to keep it as simple as possible, and only add new things if they are absolutely necessary.